### PR TITLE
Add concrete error types to JetStream

### PR DESCRIPTION
**Pull request by Jarema**
TODO:
- [x] add test for nonexisting jetstream error
- [x] check if JetStreamError is literal tuple in enum where sensible
- [x] check if what is StreamsError can be JetstreamRequestError (aggregating both request errors and response errors)
- [x] self-review

### CODE REVIEW

File: async-nats/src/jetstream/errors.rs [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+    /// Consumer filter in filter_subjects cannot be empty
+    pub const CONSUMER_EMPTY_FILTER: ErrorCode = ErrorCode(10139);
+
+    /// Duplicate source configuration detected
+    pub const SOURCE_DUPLICATE_DETECTED: ErrorCode = ErrorCode(10140);
+}
+
+/// `Error` type returned from an API response when an error occurs.
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct Error {
```
Comment (caspervonb) [inline async-nats/src/jetstream/errors.rs:442]: Move this into `jetstream::response`?

File: async-nats/src/jetstream/kv/mod.rs [hunk 2]
```diff
@@ -315,11 +315,11 @@ impl Store {
                         Some((message.message, operation, sequence, created))
                     }
                     Err(err) => {
-                        let e: std::io::Error = *err.downcast().unwrap();
-                        if e.kind() == ErrorKind::NotFound {
+                        println!("ERROR FOR DIRECT GET: {:?}", err);
```
Comment (caspervonb) [inline async-nats/src/jetstream/kv/mod.rs:318]: Take out this println

```suggestion
```

File: async-nats/src/jetstream/mod.rs [hunk 3]
```diff
@@ -132,6 +132,7 @@ use crate::Client;
 pub mod account;
 pub mod consumer;
 pub mod context;
+pub mod errors;
```
Comment (caspervonb) [inline async-nats/src/jetstream/mod.rs:135]: Think this module can be flattened into response (with into_future requests, we'll probably be having a request module too).

```suggestion
pub mod response;
```

File: async-nats/src/jetstream/errors.rs [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+
+impl ErrorCode {
+    /// Peer not a member
+    pub const CLUSTER_PEER_NOT_MEMBER: ErrorCode = ErrorCode(10040);
+
+    /// Consumer expected to be ephemeral but detected a durable name set in subject
+    pub const CONSUMER_EPHEMERAL_WITH_DURABLE: ErrorCode = ErrorCode(10019);
+
+    /// Stream external delivery prefix overlaps with stream subject
+    pub const STREAM_EXTERNAL_DEL_PREFIX_OVERLAPS: ErrorCode = ErrorCode(10022);
```
Comment (caspervonb) [inline async-nats/src/jetstream/errors.rs:29]: Consider

```suggestion
    pub const STREAM_EXTERNAL_DELETE_PREFIX_OVERLAPS: ErrorCode = ErrorCode(10022);
```

File: async-nats/src/jetstream/errors.rs [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+    pub const SEQUENCE_NOT_FOUND: ErrorCode = ErrorCode(10043);
+
+    /// Stream mirror configuration can not be updated
+    pub const STREAM_MIRROR_NOT_UPDATABLE: ErrorCode = ErrorCode(10055);
+
+    /// Expected stream sequence does not match
+    pub const STREAM_SEQUENCE_NOT_MATCH: ErrorCode = ErrorCode(10063);
+
+    /// Wrong last msg id
+    pub const STREAM_WRONG_LAST_MSG_ID: ErrorCode = ErrorCode(10070);
```
Comment (caspervonb) [inline async-nats/src/jetstream/errors.rs:128]: Consider

```suggestion
    pub const STREAM_WRONG_LAST_MESSAGE_ID: ErrorCode = ErrorCode(10070);
```

File: async-nats/Cargo.toml [hunk 6]
```diff
@@ -22,7 +22,7 @@ once_cell = "1.17.1"
 regex = "1.7.1"
 serde = { version = "1.0.152", features = ["derive"] }
 serde_json = "1.0.93"
-serde_repr = "0.1.10"
+serde_repr = "0.1.11"
```
Comment (caspervonb) [inline async-nats/Cargo.toml:25]: Intentional bump?

```suggestion
```

**Review by caspervonb — COMMENTED** [review]
Good work, some tiny nits / callouts

File: async-nats/src/jetstream/errors.rs [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+    pub const CONSUMER_PULL_WITH_RATE_LIMIT: ErrorCode = ErrorCode(10086);
+
+    /// Consumer max waiting needs to be positive
+    pub const CONSUMER_MAX_WAITING_NEGATIVE: ErrorCode = ErrorCode(10087);
+
+    /// Consumer idle heartbeat requires a push based consumer
+    pub const CONSUMER_HEARTBEAT_REQUIRES_PUSH: ErrorCode = ErrorCode(10088);
+
+    /// Consumer flow control requires a push based consumer
+    pub const CONSUMER_FC_REQUIRES_PUSH: ErrorCode = ErrorCode(10089);
```
Comment (caspervonb) [inline async-nats/src/jetstream/errors.rs:284]: Maybe

```suggestion
    pub const CONSUMER_FLOW_CONTROL_REQUIRES_PUSH: ErrorCode = ErrorCode(10089);
```

File: async-nats/src/jetstream/errors.rs [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+    pub const CLUSTER_REQUIRED: ErrorCode = ErrorCode(10010);
+
+    /// Consumer expected to be durable but a durable name was not set
+    pub const CONSUMER_DURABLE_NAME_NOT_SET: ErrorCode = ErrorCode(10018);
+
+    /// Maximum number of streams reached
+    pub const MAXIMUM_STREAMS_LIMIT: ErrorCode = ErrorCode(10027);
+
+    /// Stream mirrors can not have both start seq and start time configured
+    pub const MIRROR_WITH_START_SEQ_AND_TIME: ErrorCode = ErrorCode(10032);
```
Comment (caspervonb) [inline async-nats/src/jetstream/errors.rs:200]: Maybe

```suggestion
    pub const MIRROR_WITH_START_SEQUENCE_AND_TIME: ErrorCode = ErrorCode(10032);
```

**Review by caspervonb — CHANGES_REQUESTED** [review]
Two more abbrevs

File: async-nats/src/jetstream/errors.rs [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{error, fmt};
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Deserialize, Serialize)]
+pub struct ErrorCode(u64);
```
[the remark this answers is not in the document]
Comment (Jarema) [reply async-nats/src/jetstream/errors.rs:19]: as agreed, for now, made `jetstream::error` module not public ,and exposed Error and ErrorCode

File: async-nats/src/jetstream/mod.rs (hunk 3 above)
Comment (Jarema) [reply async-nats/src/jetstream/mod.rs]: Flattened as agreed on chat.
Error accessible via `async_nats::jetstream::Error`

**Pull request merged**